code/model/modelanimation_segments.cpp

```diff
@@ -53,93 +53,203 @@ namespace animation {
+		m_duration[pmi->id] = m_time;
 	};
 
+	ModelAnimationParseHelper::Registrar ModelAnimationSegmentWait::reg("$Wait:", &parser);
+	std::shared_ptr<ModelAnimationSegment> ModelAnimationSegmentWait::parser(ModelAnimationParseHelper* /*data*/) {
+		required_string("+Time:");
+		float time = 0.0f;
+		stuff_float(&time);
+		auto segment = std::shared_ptr<ModelAnimationSegmentWait>(new ModelAnimationSegmentWait(time));
 
-	ModelAnimationSegmentSetPHB::ModelAnimationSegmentSetPHB(const angles& angle, bool isAngleRelative) :
-		m_targetAngle(angle), m_isAngleRelative(isAngleRelative) { }
+		return segment;
+	}
+
+
+	ModelAnimationSegmentSetPHB::ModelAnimationSegmentSetPHB(std::shared_ptr<ModelAnimationSubmodel> submodel, const angles& angle, bool isAngleRelative) :
+		m_submodel(std::move(submodel)), m_targetAngle(angle), m_isAngleRelative(isAngleRelative) { }
+
+	ModelAnimationSegment* ModelAnimationSegmentSetPHB::copy() const {
+		return new ModelAnimationSegmentSetPHB(*this);
+	}
 
-	void ModelAnimationSegmentSetPHB::recalculate(const submodel_instance* /*submodel_instance*/, const bsp_info* /*submodel*/, const ModelAnimationData<>& base, int pmi_id) {
+	void ModelAnimationSegmentSetPHB::recalculate(ModelAnimationSubmodelBuffer& base, polymodel_instance* pmi) {
+		int pmi_id = pmi->id;
 		if (m_isAngleRelative) {
 			vm_angles_2_matrix(&m_instances[pmi_id].rot, &m_targetAngle);
 		}
 		else {
 			//In Absolute mode we need to undo the previously applied rotation to make sure we actually end up at the target rotation despite having only a delta we output, as opposed to just overwriting the value
 			matrix unrotate, target;
-			vm_copy_transpose(&unrotate, &base.orientation);
+			const ModelAnimationData<>& submodel = base[m_submodel].first;
+
+			vm_copy_transpose(&unrotate, &submodel.orientation);
 			vm_angles_2_matrix(&target, &m_targetAngle);
 			vm_matrix_x_matrix(&m_instances[pmi_id].rot, &target, &unrotate);
 		}
 
 		m_duration[pmi_id] = 0.0f;
 	}
 
-	ModelAnimationData<true> ModelAnimationSegmentSetPHB::calculateAnimation(const ModelAnimationData<>& /*base*/, float /*time*/, int pmi_id) const {
+	void ModelAnimationSegmentSetPHB::calculateAnimation(ModelAnimationSubmodelBuffer& base, float /*time*/, int pmi_id) const {
 		ModelAnimationData<true> data;
 		data.orientation = m_instances.at(pmi_id).rot;
-		return data;
+
+		base[m_submodel].first.applyDelta(data);
+		base[m_submodel].second = true;
+	}
+
+	void ModelAnimationSegmentSetPHB::exchangeSubmodelPointers(ModelAnimationSet& replaceWith) {
+		m_submodel = replaceWith.getSubmodel(m_submodel);
+	}
+
+	ModelAnimationParseHelper::Registrar ModelAnimationSegmentSetPHB::reg("$Set Orientation:", &parser);
+	std::shared_ptr<ModelAnimationSegment> ModelAnimationSegmentSetPHB::parser(ModelAnimationParseHelper* data) {
+		angles angle;
+		bool isRelative = true;
+
+		required_string("+Angle:");
+		stuff_angles_deg_phb(&angle);
+		isRelative &= !optional_string("+Absolute");
```

Comment:
I think it would be nice to allow the modder to specify either `+Absolute` or `+Relative` here